csharp/ql/src/Language Abuse/UselessUpcast.ql

```diff
@@ -56,47 +92,74 @@ class ExplicitUpcast extends ExplicitCast {
+    exists(Call c, InstanceCallable target, ValueOrRefType t |
       this.isArgument(c, target) |
+      t = c.(QualifiableExpr).getQualifier().getType() and
+      hasInstanceCallable(t, other, target.getName()) and
+      args = c.getNumberOfArguments() and
+      other != target
+    )
+  }
+
+  /** Holds if this upcast may be used to disambiguate the target of an extension method call. */
+  pragma [nomagic]
+  private predicate isDisambiguatingExtensionCall(ExtensionMethod other, int args) {
+    exists(ExtensionMethodCall c, ExtensionMethod target, ValueOrRefType t |
+      this.isArgument(c, target) |
+      not c.isOrdinaryStaticCall() and
+      t = target.getParameter(0).getType() and
+      hasExtensionMethod(t, other, target.getName()) and
+      args = c.getNumberOfArguments() and
+      other != target
+    )
+  }
+
+  pragma [nomagic]
+  private predicate isDisambiguatingStaticCall0(StaticCall c, StaticCallable target, ValueOrRefType t) {
+    this.isArgument(c, target) and
+    (
       t = c.(QualifiableExpr).getQualifier().getType()
       or
-      not exists(c.(QualifiableExpr).getQualifier()) and
+      not c.(QualifiableExpr).hasQualifier() and
       t = target.getDeclaringType()
     )
   }
 
-  /**
-   * Holds if this upcast may be used to affect call resolution in a non-extension
-   * method call.
-   */
-  private predicate isDisambiguatingNonExtensionMethodCall() {
-    exists(NonExtensionMethod target, NonExtensionMethod other, ValueOrRefType t |
-      this.isDisambiguatingNonExtensionMethod0(target, t) |
-      hasCallable(t, other, target.getName()) and
+  /** Holds if this upcast may be used to disambiguate the target of a static call. */
+  pragma [nomagic]
+  private predicate isDisambiguatingStaticCall(StaticCallable other, int args) {
+    exists(StaticCall c, StaticCallable target, ValueOrRefType t |
+      this.isDisambiguatingStaticCall0(c, target, t) |
+      hasStaticCallable(t, other, target.getName()) and
+      args = c.getNumberOfArguments() and
       other != target
     )
   }
 
-  /**
-   * Holds if this upcast may be used to affect call resolution in an extension
-   * method call.
-   */
-  private predicate isDisambiguatingExtensionMethodCall() {
-    exists(Call c, ExtensionMethod target, ExtensionMethod other, ValueOrRefType t |
-      this.isArgument(c, target) |
-      t = c.getArgument(0).getType() and
-      hasExtensionMethod(t, other, target.getName()) and
-      other != target
+  /** Holds if this upcast may be used to disambiguate the target of a call. */
+  private predicate isDisambiguatingCall() {
+    exists(Callable other, int args |
+      this.isDisambiguatingInstanceCall(other, args)
+      or
+      this.isDisambiguatingExtensionCall(other, args)
+      or
+      this.isDisambiguatingStaticCall(other, args)
+      |
+      args >= getMinimumArguments(other) and
+      (
+        not exists(getMaximumArguments(other))
+        or
+        args <= getMaximumArguments(other)
```

Comment:
Combine both cases into `not args > getMaximumArguments(other)`

Reply:
Yes, that is better, thanks!